### PR TITLE
Changes so File object creation doesn't violate on of the MUSTs

### PR DESCRIPTION
**Pull request by rpiazza**
Added three new exceptions: DependentPropertiestError, AtLeastOnePropertyError, MutuallyExclusivePropertiesError
Added tests for NetworkTraffic, Process, URL, WindowsRegistryKey and X509Certificate
Added error tests for EmailMessage, NetworkTraffic, Artifact,
Added interproperty checker methods to the base class:  _check_mutually_exclusive_properties, _check_at_least_one_property and _check_properties_dependency
Added interproperty checkers to Artifact, EmailMIMEComponent, EmailMessage, NetworkTraffic
Made NetworkTraffic.protocols required
Added X509V3ExtenstionsType class
Use EmbeddedObjectProperty for X509Certificate.x509_v3_extensions

### CODE REVIEW

File: stix2/observables.py [hunk 1]
```diff
@@ -97,6 +105,11 @@ class EmailMessage(_Observable):
         'raw_email_ref': ObjectReferenceProperty(),
     }
 
+    def _check_object_constaints(self):
+        super(EmailMessage, self)._check_object_constaints()
+        self._check_properties_dependency(["is_multipart"], ["body_multipart"])
+        # self._dependency(["is_multipart"], ["body"], [False])
```
Comment (clenk) [inline stix2/observables.py]: Is this commented-out line still being worked on?

File: stix2/base.py [hunk 2]
```diff
@@ -47,6 +47,36 @@ def _check_property(self, prop_name, prop, kwargs):
...
                 raise InvalidValueError(self.__class__, prop_name, reason=str(exc))
 
+    # interproperty constraint methods
+
+    def _check_mutually_exclusive_properties(self, list_of_properties, at_least_one=True):
+        count = 0
+        current_properties = self.properties_populated()
+        for x in list_of_properties:
+            if x in current_properties:
+                count += 1
```
Comment (rpiazza) [inline stix2/base.py]: looks good.

File: stix2/observables.py (hunk 1 above)
Comment (rpiazza) [inline stix2/observables.py]: Yes - it doesn't work, but there is such a dependency

File: stix2/observables.py [hunk 3]
```diff
@@ -315,5 +347,5 @@ class X509Certificate(_Observable):
         'subject_public_key_algorithm': StringProperty(),
         'subject_public_key_modulus': StringProperty(),
         'subject_public_key_exponent': IntegerProperty(),
-        'x509_v3_extensions': Property(),
+        'x509_v3_extensions': EmbeddedObjectProperty(type=X509V3ExtenstionsType),
```
Comment (rpiazza) [inline stix2/observables.py]: Yes there are a few of those - mostly concerning extensions.  I will add them with the next changes.

**Pull request merged**